enterprise/backend/src/metabase_enterprise/dependencies/models/analysis_finding.clj

```diff
@@ -0,0 +1,63 @@
+(ns metabase-enterprise.dependencies.models.analysis-finding
+  (:require
+   [metabase.lib.normalize :as lib.normalize]
+   [metabase.models.interface :as mi]
+   [methodical.core :as methodical]
+   [toucan2.core :as t2]))
+
+(methodical/defmethod t2/table-name :model/AnalysisFinding [_model] :analysis_finding)
+
+(derive :model/AnalysisFinding :metabase/model)
+
+(defn- normalize-analysis-finding [finding]
+  (some->> (mi/json-out-with-keywordization finding)
+           (into #{} (map #(lib.normalize/normalize :metabase.lib.schema.validate/error %)))))
+
+(t2/deftransforms :model/AnalysisFinding
+  {:analyzed_entity_type mi/transform-keyword
+   :finding_details {:in  mi/json-in
+                     :out normalize-analysis-finding}})
+
+;; This generally shouldn't be rebound in real code, but making it dynamic is convenient for testing
+(def ^:dynamic *current-analysis-version*
+  "Current version of the analysis logic.
+  This should be incremented when the analysis logic changes.
+  The background task will re-analyze anything with out-of-date analyses."
+  3)
+
+(defn upsert-analysis!
+  "Given the details of an AnalysisFinding row, upsert the data into the actual db."
+  [type instance-id result finding-details]
+  (let [update {:analyzed_at (mi/now)
+                :analysis_version *current-analysis-version*
+                :result result
+                :finding_details finding-details}
+        existing-id (t2/select-one-fn :id [:model/AnalysisFinding :id]
+                                      :analyzed_entity_type type
+                                      :analyzed_entity_id instance-id)]
+    (if existing-id
+      (t2/update! :model/AnalysisFinding existing-id update)
+      (t2/insert! :model/AnalysisFinding
+                  (assoc update
+                         :analyzed_entity_type type
+                         :analyzed_entity_id instance-id)))))
+
+(def ^:private table-info
+  {:card [:model/Card :report_card :report_card.* :report_card.id]
```

Comment:
Kinda confusing to structure the data this way, you have to go look at the function below to even figure out what's going on here. This is a good place to use maps with keys.

(On that note you literally only need to know the Toucan model anyways since all the other stuff can be looked up or derived from that)